src/tactic/tcache/hijack.lean

```diff
@@ -0,0 +1,53 @@
+import .tactic
+
+namespace tcache
+
+open tactic
+
+section locking
+
+@[inline] def LOCKED_FLAG : name := `tcache.locked
+
+@[inline] meta def is_locked : tactic bool :=
+do o ← get_options,
+   return $ o.get_bool LOCKED_FLAG ff
+
+@[inline] meta def set_locked (b : bool) : tactic unit :=
+do o ← get_options,
+   set_options $ o.set_bool LOCKED_FLAG b
+
+@[inline] meta def lock : tactic unit := set_locked tt
+
+@[inline] meta def unlock : tactic unit := set_locked ff
+
+@[inline] meta def try_under_lock {α : Type} (alt : tactic α) (t : tactic α) : tactic α := do
+  b ← is_locked,
+  if b then alt
+  else do
+    lock,
+    r ← t,
+    unlock,
+    return r
+
+@[inline] meta def try_apply_under_lock {α : Type} (h : tactic α → tactic α) (t : tactic α) : tactic α :=
+try_under_lock t (h t)
+
+@[inline] meta def ensure_no_caching {α : Type} (t : tactic α) : tactic α := do
+  b ← is_locked,
+  lock,
+  r ← t,
+  set_locked b,
+  return r
+
+end locking
+
+section hook
+
+@[inline] meta def bind {α β : Type} (t₁ : tactic α) (t₂ : α → tactic β) : tactic β :=
+try_apply_under_lock tcache_force $ interaction_monad_bind t₁ t₂
+
+@[priority 2000, inline] meta instance has_bind : has_bind tactic := ⟨@bind⟩
```

Comment:
Oooh, this is a nice trick.

Reply:
In community lean there is a much better way to do this now. You use `interactive.executor` (https://github.com/leanprover-community/lean/blob/a148f386e748551502792882a5507577dbee9b14/library/init/meta/tactic.lean#L100) which I added ages ago to take control only over `begin ... end` blocks and `by ...` blocks, and without the options trickery which appears at the top of this file (since we aren't doing something dodgy in the new version, just providing a higher priority tactic block implementation which---much more sensibly---is only called once).

Reply:
(Perhaps no-one has noticed, but the code for `begin ... end` blocks generated by community lean invokes the highest priority `executor` using the standard typeclass resolution mechanism.)

Reply:
One other point: The real reason I wasn't happy with intercepting `begin ... end`/`by ...` and not `has_bind` was not the hackery of this approach (preventing recursively caching on every bind), but the fact that intercepting `has_bind` is like a cancer, which you cannot allow to spread to any of your actual `tactic` programs! Since then, normal function calls in `meta` land from a `do` block will be cached! The result is that you couldn't write tactic code in any file which indirectly included `tactic.tcache.enable`, which was obviously unacceptable. (I realized this late, and made the `executor` stuff after.)

Reply:
How hard is it to update this PR so it uses the `interactive.executor` machinery, now that mathlib is running on community Lean?

Reply:
It essentially involves deleting files from this branch. I'll give it a go.